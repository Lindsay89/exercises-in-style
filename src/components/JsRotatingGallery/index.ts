export { default as JsRotatingGallery } from "./JsRotatingGallery";
