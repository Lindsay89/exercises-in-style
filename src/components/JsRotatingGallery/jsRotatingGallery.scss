.eis-jsRotatingGallery {
  padding: 2rem;
  background-color: blue;
  height: 100%;
  overflow: hidden;

  .jsRotatingGallery-container {
    position: relative;
    height: 100%;
    align-items: center;
    display: flex;
    justify-content: center;
    transform-style: preserve-3d;
    animation: rotating 20s linear infinite;
    /* animation-play-state: paused;
    transform: rotateY(0deg); */
    /* 
    &.animation-stop {
      animation: none;
    }
    &.animation-active {
      animation-play-state: running;
    } */
    .jsRotatingGallery-content {
      transform-style: preserve-3d;
      height: 100%;
      align-items: center;
      display: flex;
      justify-content: center;
      width: 100%;
      transition: all;
      transition-delay: 0.5s;
      transition-duration: 1s;
      .jsRotatingGallery-image {
        position: absolute;
        transform-origin: center;
        width: 30%;
        height: 50%;

        img {
          /* position: absolute;
          top: 0;
          left: 0; */
          width: 100%;
          height: 100%;
          object-fit: cover;
        }
      }
    }
  }
}
@keyframes rotating {
  from {
    transform: rotateY(0deg);
  }
  to {
    transform: rotateY(360deg);
  }
}
