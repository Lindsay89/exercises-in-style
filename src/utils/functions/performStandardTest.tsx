import React from "react";
import { render } from "@testing-library/react";
import { DefaultProps } from "../types/defaultProps";

const performStandardTest = (
  Component: React.FunctionComponent<any>,
  defaultProps: DefaultProps,
  defaultClassName: string
) => {
  const dataTestId = defaultProps["data-testid"] as string;
  it("should render without explode", () => {
    const { getByTestId } = render(<Component {...defaultProps} />);

    expect(getByTestId(dataTestId).id).toMatchSnapshot();
  });

  it('should accept an "id" prop', () => {
    const { getByTestId } = render(<Component {...defaultProps} id="foo-id" />);

    expect(getByTestId(dataTestId).id).toBe("foo-id");
  });

  it("should have default className", () => {
    const { getByTestId } = render(<Component {...defaultProps} />);

    expect(getByTestId(dataTestId).classList?.contains(defaultClassName)).toBe(
      true
    );
  });

  it("should allow adding custom classes", () => {
    const { getByTestId } = render(
      <Component {...defaultProps} className="bar" />
    );

    expect(getByTestId(dataTestId).classList.contains("bar")).toBe(true);
  });

  it("should allow to define custom style", () => {
    const { getByTestId } = render(
      <Component {...defaultProps} style={{ margin: "30px" }} />
    );

    expect(getByTestId(dataTestId).getAttribute("style")).toContain("margin");
  });
};
export default performStandardTest;
