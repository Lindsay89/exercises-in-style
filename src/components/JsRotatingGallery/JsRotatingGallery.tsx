import React, { MouseEventHandler, useEffect, useRef } from "react";
import classNames from "classnames";
import { DefaultProps } from "../../utils/types";

// styles
import "./jsRotatingGallery.scss";

// Required Props
interface JsRotatingGalleryRequiredProps {
  // An array of images to show
  images: string[];
}

// Optional Props
interface JsRotatingGalleryOptionalProps extends DefaultProps {}

// Combined required and optional props to build the full prop interface
export interface JsRotatingGalleryProps
  extends JsRotatingGalleryRequiredProps,
    JsRotatingGalleryOptionalProps {}

// use the optional prop interface to define the default props
const defaultProps: JsRotatingGalleryOptionalProps = {
  "data-testid": "eis-jsRotatingGallery",
};

const JsRotatingGallery: React.FC<JsRotatingGalleryProps> = ({
  className,
  images,
  ...rest
}) => {
  const classList = classNames("eis-jsRotatingGallery", className);
  let interval = useRef();

  useEffect(() => {
    const images = document.getElementsByClassName(
      "jsRotatingGallery-image"
    ) as HTMLCollectionOf<HTMLImageElement>;
    const container = document.getElementsByClassName(
      "jsRotatingGallery-container"
    )[0];
    /* container.style.setProperty('animation-play-state','running') */
    container.classList.add("animation-active");
    const { width } = container.getBoundingClientRect();
    for (let i = 0; i < images.length; i += 1) {
      const rotateValueCalc = (360 / images.length) * i;
      const rotateValue = ` rotateY(${rotateValueCalc}deg)`;
      const transformValue = ` ${rotateValue} translateZ(${width / 2}px)`;
      images[i].style.setProperty("transform", transformValue);
    }
    interval.current = setInterval(() => console.log("ciao"), 2000);
    container.classList.add("rotation-animation");
  }, []);

  const handleOnImgMouseEnter = (e, index) => {
    console.log(index);
    const content = document.getElementsByClassName(
      "jsRotatingGallery-content"
    )[0];
    const container = document.getElementsByClassName(
      "jsRotatingGallery-container"
    )[0];
    const rotateValueCalc = 360 - (360 / images.length) * index;
    clearInterval(interval.current);
    container.classList.add("animation-stop");
    document
      .getElementsByClassName("jsRotatingGallery-content")[0]
      .style.setProperty("transform", `rotateY(${rotateValueCalc}deg)`);
    e.target.style.setProperty("transform", "scale(2)");
    document
      .getElementsByClassName("jsRotatingGallery-container")[0]
      .style.setProperty("animation-play-state", "paused");
    /* 
    e.stopPropagation();
    e.preventDefault(); */
  };
  const handleOnMouseLeave = (e, index) => {
    /* document
      .getElementsByClassName("jsRotatingGallery-container")[0]
      .style.setProperty("animation-play-state", "running"); */
    document
      .getElementsByClassName("jsRotatingGallery-container")[0]
      .classList.remove("animation-stop");
    document
      .getElementsByClassName("jsRotatingGallery-container")[0]
      .style.setProperty("animation-play-state", "running");
    e.target.style.setProperty("transform", "scale(1)");
    e.stopPropagation();
    e.preventDefault();
  };
  return (
    <div className={classList} {...rest}>
      <div className="jsRotatingGallery-container">
        <div className="jsRotatingGallery-content">
          {images.map((img, index) => (
            <span
              key={img}
              style={{ "--imgPosition": index } as React.CSSProperties}
              className="jsRotatingGallery-image"
              onMouseEnter={(e) => handleOnImgMouseEnter(e, index)}
              onMouseLeave={(e) => handleOnMouseLeave(e, index)}
            >
              <img src={img} alt={img} />
            </span>
          ))}
        </div>
      </div>
    </div>
  );
};

JsRotatingGallery.defaultProps =
  defaultProps as Partial<JsRotatingGalleryOptionalProps>;

export default React.memo(JsRotatingGallery);
