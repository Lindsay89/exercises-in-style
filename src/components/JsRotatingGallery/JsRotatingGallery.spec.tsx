import React from "react";
import JsRotatingGallery from "./JsRotatingGallery";
import performStandardTest from "../../utils/functions/performStandardTest";

const defaultProps = {
  "data-testid": "eis-jsRotatingGallery",
  images: [
    "https://picsum.photos/1024/600?random&category=nature",
    "https://picsum.photos/1024/600?random&category=sun",
  ],
};

describe("JsRotatingGallery Component", () => {
  performStandardTest(JsRotatingGallery, defaultProps, "eis-jsRotatingGallery");
});
