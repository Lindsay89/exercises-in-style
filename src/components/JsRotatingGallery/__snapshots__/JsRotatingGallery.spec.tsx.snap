// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsRotatingGallery Component should render without explode 1`] = `""`;
